zookeeper-server/src/main/java/org/apache/zookeeper/MultiTransactionRecord.java

```diff
@@ -28,20 +28,25 @@
 import java.util.List;
 
 /**
- * Encodes a composite transaction.  In the wire format, each transaction
+ * Encodes a composite operation.  In the wire format, each operation
  * consists of a single MultiHeader followed by the appropriate request.
  * Each of these MultiHeaders has a type which indicates
- * the type of the following transaction or a negative number if no more transactions
+ * the type of the following operation or a negative number if no more operations
  * are included.
+ * All of the operations must be from the same OpKind.
+ * Note: This class is called MultiTransactionRecord because of legacy reasons,
+ * MultiOperationRecord would be more accurate.
  */
 public class MultiTransactionRecord implements Record, Iterable<Op> {
```

Comment:
Should we create a new MultiReadRecord instead of reusing this one? Since the read is not txn, and we don't allow read mixed with write.

It avoided some unnecessary check in the code, and also less confusing to user.

Reply:
On one hand, I definitely agree with you that the name is confusing and could be improved. Going into implementation details, I think it would be the best to create a `MultiOperationRecord` base (super) class which could have `MultiReadRecord` and `MultiTransactionRecord` subclasses.
On the other hand, (besides resolving confusion caused by the name) I do not see how it helps as to avoid any checks. It would require us to overload some function for Txn/Read operations which would lead to more duplicated code. Another way is that we could have an `instanceof` check at the beginning of these function instead of the `OpKind` checks but again, this would result the some checks.
Since `Op`, `OpResult,` and `MultiResponse` is preserved as a single class (though these could be extended into more subclasses as well like `ReadOp/TransactionOp`, `MultiTxnResponse/MultiReadResponse`) to maintain the current workflow, I say it is better and more consistent to preserve it as well and make decisions based on `OpKind`.
But of course, I am convincible on this topic so feel free to argue! :) What do you think?

Reply:
One more note to this topic:
As I am a fresh contributor don't know how important the backward compatibility of a class like this is. I am not sure whether `MultiTransactionRecord` is even used outside this library since the multi functions do not require the user to construct it. So, maybe a simple rename to `MultiOperationRecord` could resolve the confusion? Would it be a deal breaker?

Reply:
I feel we can change the name of classes as it is not a public API and class names are not stored on snapshots/ txlog